.gitlab-ci.yml

```diff
@@ -286,147 +206,149 @@ build-core:
     - docker
   rules:
     - changes:
-        - dockerfiles/**/*
+        - docker/**/*
         - scripts/**/*
-        - common/**/*
         - core/**/*
+        - common/**/*
         - CMakeLists.txt
         - cmake/**/*
         - .gitlab-ci.yml
       when: on_success
-    - exists:
-      - "build-core-base"
-      when: on_success
   script:
     - BRANCH_LOWER=$(echo "$CI_COMMIT_REF_NAME" | tr '[:upper:]' '[:lower:]')
     - echo "$BRANCH_LOWER"
-    - env
     - ./scripts/generate_datafed.sh
-    - cat ./config/datafed.sh
     - docker login "${REGISTRY}" -u "${HARBOR_USER}" -p "${HARBOR_DATAFED_GITLAB_CI_REGISTRY}"
-    - docker build --build-arg BASE_CORE_IMAGE="${REGISTRY}/${IMAGE_TAG}-base-${BRANCH_LOWER}:latest" -f core/docker/Dockerfile.core.ubuntu -t "${REGISTRY}/${IMAGE_TAG}-${BRANCH_LOWER}" .
+    - docker build --build-arg DEPENDENCIES="${REGISTRY}/datafed/dependencies-${BRANCH_LOWER}:latest" --build-arg RUNTIME="${REGISTRY}/datafed/runtime-${BRANCH_LOWER}:latest" -f core/docker/Dockerfile -t "${REGISTRY}/${IMAGE_TAG}-${BRANCH_LOWER}:latest" .
     - docker push "${REGISTRY}/${IMAGE_TAG}-${BRANCH_LOWER}"
 
-build-foxx:
+build-repo:
+  needs: ["build-dependencies", "build-runtime"]
   stage: build
-  needs: ["signal"]
   variables:
+    IMAGE_TAG: "datafed/repo"
     GIT_STRATEGY: clone
-    DATAFED_DATABASE_PASSWORD: "${CI_DATAFED_DATABASE_PASSWORD}"
-    DATAFED_ZEROMQ_SYSTEM_SECRET: "${CI_DATAFED_ZEROMQ_SYSTEM_SECRET}"
   tags:
-    - ci-datafed-arango
+    - ci-datafed-repo
+    - docker
+  rules:
+    - changes:
+        - docker/**/*
+        - scripts/**/*
+        - web/**/*
+        - common/proto/**/*
+        - .gitlab-ci.yml
+      when: on_success
```

Comment:
Another thing you could have pointed out, there is probably room for optimization around the rules. Did you look at these in any detail?